src/Dialect/ONNX/ONNXShapeHelper.hpp

```diff
@@ -0,0 +1,42 @@
+//===----------------ONNXShapeHelper.hpp - help for shapes----------------=== //
+//
+// Copyright 2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file has the computations to compute the shapes using the new index expr
+// approach.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Support/LogicalResult.h"
+
+#include "src/Dialect/ONNX/IndexExpr.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+
+using namespace mlir;
+
+size_t ArrayAttrSize(ArrayAttr a);
+size_t ArrayAttrSize(Optional<ArrayAttr> a);
+int64_t ArrayAttrIntVal(ArrayAttr a, int i);
+int64_t ArrayAttrIntVal(Optional<ArrayAttr> a, int i);
+// Returns the ConstantOp which defines an MLIR Value or null.
+ONNXConstantOp getONNXConstantOp(Value value);
+
+DenseElementsAttr getDenseElementAttributeFromValue(Value value);
+
+bool getIntegerLiteralFromValue(Value value, int64_t &intLit);
+
+LogicalResult HandleSliceOpParams(ONNXSliceOp *sliceOp,
```

Comment:
What is shared between shape inference and lowering is the outputDims. The other IndexExprs are auxiliary expressions that are used only in lowering. Can we have a uniform interface across the operations?  Make the OutputDim the first SmallVectorImpl and others with default NULL.

Reply:
I changed the interface in the last edition to hide the long list of parameters inside a struct. To respond to your question: during shape inference, only outputDims are needed; the other one are used to compute the outputDim. Here is the new interface
```
/// When defining support for a new op, add one such stuct which must minimally
/// compute the outputDims present in the parent class. Computation should be
/// performed using a `Compute` function. Return success on successful
/// computation of all the IndexExpr. During shape inference, object is built
/// using a null-ptr rewriter; during lowering, the rewriter is nonnull and will
/// be used to generate code.
template <class OP>
struct ONNXOpShapeHelper {
  ONNXOpShapeHelper(OP *newOp, ConversionPatternRewriter *rewriter);

  // Define in every children.
  LogicalResult Compute(ONNXSliceOpAdaptor operandAdaptor) {
    llvm_unreachable("implement in child structs");
  }

  // Data that must be present for every ShapeHelper operation. Op and context
  // are initialized in the constructor, and outputDims is computed by the
  // child's struct `Compute` function.
  OP *op;
  IndexExprContext context;
  SmallVector<IndexExpr, 4> outputDims;
};

struct ONNXSliceOpShapeHelper : public ONNXOpShapeHelper<ONNXSliceOp> {
  ONNXSliceOpShapeHelper(
      ONNXSliceOp *newOp, ConversionPatternRewriter *rewriter);

  LogicalResult Compute(ONNXSliceOpAdaptor operandAdaptor);

  // Additional data for SliceOp.
  SmallVector<IndexExpr, 4> starts;
  SmallVector<IndexExpr, 4> ends;
  SmallVector<IndexExpr, 4> steps;
};
```

What is in `ONNXOpShapeHelper` is the minimum that is needed for shape inference. When generating a subclass for a specific op, we can add additional parameters, but during the shape inference, only the first set is needed.

Inside the IndexExpression, the difference between shape inference and lowering is that during shape inference, expressions are either constant or '?'. We cannot/do not want to gen code during shape inference. so we will not go beyond the '?'. Note the expressions can go back to constant

res = (1 > 0) ? 3 : runtime-expo

will result in a constant...

During lowering, we will generate code.

Note that in the interface above, the `Compute(...)` function can detect if we are in shape inference or not by using the  `context.IsShapeInferencePass()` test and thus can customize the computations, if desired.

Bottom line, during inference: must compute `outputDims`. During lowering, compute that one plus any others needed for lowering. Hope this helps.